### PR TITLE
use dbt to test source and output data

### PR DESCRIPTION
**Pull request by damonmcc**
## why

related to issue #116

We've talked about testing source and output data for our builds. `dbt` is often used to model, transform, and test data. By only adding the modeling and testing parts, we can use `dbt` for data validation without refactoring product build logic.

Template DB seems like a good product to try this out on.

## changes

- python package changes from adding `dbt-postgres` to `requirements.in` and compiling via github action
  - pydantic must be rolled back to use `dbt-core>1.5` (for now, `pydantic 2.x` compatibility seems to be coming soon)
- add dbt project configuration for Template DB
- add yml files to model the output tables and a single source table of Template DB
- add tests to those models using built-in and `dbt-utils` tests
- run dbt tests at the end of `python3 -m build_scripts.transform`

## notes
- a test passing [here](https://github.com/NYCPlanning/data-engineering/actions/runs/7392820649/job/20111939790#step:9:158)
- a test that raising a warning [here](https://github.com/NYCPlanning/data-engineering/actions/runs/7392820649/job/20111939790#step:9:134)

### CODE REVIEW

File: products/template/build_scripts/transform.py [hunk 1]
```diff
@@ -1,16 +1,41 @@
...
         PG_CLIENT.engine_uri,
         SQL_QUERY_DIR / f"{step_query}.sql",
     )
 
 
+def _run_dbt_step(run_args=list[str]):
+    dbt_cli_call = f"dbt {' '.join(run_args)}"
+    logger.info(f"Running '{dbt_cli_call}' ...")
+    dbt_runner = dbtRunner()
+    runner_result: dbtRunnerResult = dbt_runner.invoke(run_args)
```
Comment (damonmcc) [inline products/template/build_scripts/transform.py:21]: this is where dbt is called, see docs [here](https://docs.getdbt.com/reference/programmatic-invocations)

File: products/template/dbt_project.yml
Comment (damonmcc) [inline products/template/dbt_project.yml:1]: this file configues the dbt project

File: products/template/profiles.yml
Comment (damonmcc) [inline products/template/profiles.yml:1]: this file configures database connections for dbt

File: products/template/build_scripts/transform.py (hunk 1 above)
Comment (sf-dcp) [reply products/template/build_scripts/transform.py:21]: is this command equivalent to `dbt run` CLI command?

File: products/template/build_scripts/transform.py [hunk 2]
```diff
@@ -21,12 +46,21 @@ def intermediate():
     _run_step("historic_landmarks")
 
 
-def outputs():
+def product():
```
Comment (sf-dcp) [inline products/template/build_scripts/transform.py:49]: from my limited research, it seems that it is possible to invoke table builds via one CLI command (`dbt run`). I wonder if there is a similar python way to do that...

File: products/template/profiles.yml
Comment (sf-dcp) [inline products/template/profiles.yml:1]: what does this code do? Does it configure the schema where to output tables from `product()` fn?

**Comment by sf-dcp** [comment]
Really exciting stuff!

From watching a couple of dbt tutorials, I saw a couple of interesting things that seem to be useful: 

* Having 1 SQL file per table (model) and ability to reference them via Jinja

  image


For example: during `staging` step, you create a bunch of tables such as `stg_nypl_libraries`. From my understanding, you can create a `stg_nypl_libraries.sql` model with the following content:

```SQL
SELECT
    name AS library_name,
    locality AS borough,
    wkb_geometry
FROM nypl_libraries;
```

Instead of: 

```SQL
-- "staging.sql" file

CREATE TABLE stg_nypl_libraries AS SELECT
    name AS library_name,
    locality AS borough,
    wkb_geometry
FROM nypl_libraries;
```


And they you can reference the table in other queries as `{{ ref('stg_nypl_libraries') }}`.

File: products/template/build_scripts/transform.py [hunk 3]
```diff
@@ -21,12 +46,21 @@ def intermediate():
...
+def test():
+    _run_dbt_step(["deps"])
+    _run_dbt_step(["debug"])
+    _run_dbt_step(["test", "--select", "source:*"])
+    _run_dbt_step(["test", "--select", "templatedb"])
+    _run_dbt_step(["test", "--select", "aggregation"])
 
 
 if __name__ == "__main__":
     staging()
```
Comment (sf-dcp) [inline products/template/build_scripts/transform.py:63]: Not sure if we need `staging()` if all it does is creates copies of source tables. Thoughts?

File: products/template/build_scripts/transform.py (hunk 1 above)
Comment (damonmcc) [reply products/template/build_scripts/transform.py:21]: only if run is passed as the first `run_args`. so  `_run_dbt_step(["debug"])` is equivalent to `dbt debug`

I chose to align with the `_run_step` function above but maybe this shouldn't have the word "run" in the name?

File: products/template/build_scripts/transform.py (hunk 2 above)
Comment (damonmcc) [reply products/template/build_scripts/transform.py:49]: I think it'd be `dbt_runner.invoke(["run"])` and with the wrapper I wrote above it'd be `_run_dbt_step(["run"])`

File: products/template/profiles.yml
Comment (damonmcc) [reply products/template/profiles.yml:1]: (for some reason the comment I left here is labeled as outdated)

this file is teh connection profile ([docs](https://docs.getdbt.com/docs/core/connect-data-platform/connection-profiles)). it configures database connections, including which schema to default to (I think any can be used if it's explicitely specified in a query)

File: products/template/build_scripts/transform.py (hunk 3 above)
Comment (damonmcc) [reply products/template/build_scripts/transform.py:63]: true it doesn't do much, but it does a couple of valuable things: selects the subset of columns that are actually useful and renames some to be either more readable or more standardized

see [docs](https://docs.getdbt.com/best-practices/how-we-structure/2-staging#staging-models) on ideas behind the value of using staging models

**Comment by damonmcc** [comment]
> For example: during `staging` step, you create a bunch of tables such as `stg_nypl_libraries`. From my understanding, you can create a `stg_nypl_libraries.sql` model with the following content:
> 
> ```sql
> SELECT
>     name AS library_name,
>     locality AS borough,
>     wkb_geometry
> FROM nypl_libraries;
> ```
> 
> Instead of:
> 
> ```sql
> -- "staging.sql" file
> 
> CREATE TABLE stg_nypl_libraries AS SELECT
>     name AS library_name,
>     locality AS borough,
>     wkb_geometry
> FROM nypl_libraries;
> ```
> 
> And then you can reference the table in other queries as `{{ ref('stg_nypl_libraries') }}`.

yup! I started down that route in an "experiment" PR [here](https://github.com/NYCPlanning/data-engineering/pull/344) at first. but seems nice to start to model and test products that exist and save refactoring transformations for later

File: products/template/profiles.yml
Comment (sf-dcp) [reply products/template/profiles.yml:1]: I see. +1 to dbt from me for their docs👏

File: products/template/build_scripts/transform.py (hunk 1 above)
Comment (sf-dcp) [reply products/template/build_scripts/transform.py:21]: Got it! I think "run" in the fn name is a bit confusing given that there is a command with that keyword. How about `_execute_dbt_step` or `_perform_dbt_step`?

File: products/template/build_scripts/transform.py (hunk 2 above)
Comment (sf-dcp) [reply products/template/build_scripts/transform.py:49]: I thought you were using `dbt` to build the tables so was confused having multiple steps in the `product` fn. Makes sense

**Pull request merged**